### DISCUSSION TITLE
Multithreading

### DISCUSSION THREAD

**Discussion by JplBal** [cesanta/mongoose#2085] [Ask questions, get answers] [answered]
Hello everyone
I'm using version 7.9.
I'm experimenting with the "multi threaded" example, which I use with VisualStudio 2022.
I encounter 2 problems:
1/ in windows environment, at the level of the "close(p->sock); // Close the connection" line, the "close" function is not defined for the compiler. Should I use another function please ? closesocket ?
2/ in a linux environment, I thought that the call to the close(p->sock) function would trigger the MG_EV_CLOSE event at the level of the pipe's event handler, this is not the case, which leads me to think that the pipe did not close (I dont see any corresponding traces). Is this the correct behaviour please ?
If yes, how can I close the pipe from the working thread side, please?
Thanks in advance

This is a trace capture:
1273296 3 mongoose.c:3493:mg_listen     1 0x4 http://localhost:8011
1275f05 3 mongoose.c:4377:accept_conn   2 0x5 accepted 127.0.0.1.33314 -> 127.0.0.1.8011
1275f05 3 mongoose.c:4233:read_conn     2 0x5 snd 0/0 rcv 0/2048 n=243 err=0
1275f06 2 MongooseTest.cpp:53:pcb       3 0x7 1 0x5555555863b0
1275f06 3 mongoose.c:4437:mg_mkpipe     3 0x7 pipe 6
1277cae 2 MongooseTest.cpp:53:pcb       3 0x7 2 0x5555555863b0
1278098 2 MongooseTest.cpp:53:pcb       3 0x7 2 0x5555555863b0
127847f 2 MongooseTest.cpp:53:pcb       3 0x7 2 0x5555555863b0
127847f 3 mongoose.c:4233:read_conn     3 0x7 snd 0/0 rcv 0/2048 n=2 err=4
127847f 2 MongooseTest.cpp:53:pcb       3 0x7 7 0x5555555863b0
127847f 2 MongooseTest.cpp:53:pcb       3 0x7 2 0x5555555863b0
127847f 3 mongoose.c:4244:write_conn    2 0x5 snd 65/2048 rcv 0/2048 n=65 err=4
1278869 2 MongooseTest.cpp:53:pcb       3 0x7 2 0x5555555863b0
1278c52 2 MongooseTest.cpp:53:pcb       3 0x7 2 0x5555555863b0
127903c 2 MongooseTest.cpp:53:pcb       3 0x7 2 0x5555555863b0
1279425 2 MongooseTest.cpp:53:pcb       3 0x7 2 0x5555555863b0
127980e 2 MongooseTest.cpp:53:pcb       3 0x7 2 0x5555555863b0
1279bf8 2 MongooseTest.cpp:53:pcb       3 0x7 2 0x5555555863b0
1279fe1 2 MongooseTest.cpp:53:pcb       3 0x7 2 0x5555555863b0
127a3cb 2 MongooseTest.cpp:53:pcb       3 0x7 2 0x5555555863b0
127a7b4 2 MongooseTest.cpp:53:pcb       3 0x7 2 0x5555555863b0
127ab9e 2 MongooseTest.cpp:53:pcb       3 0x7 2 0x5555555863b0
127af87 2 MongooseTest.cpp:53:pcb       3 0x7 2 0x5555555863b0
127b239 2 MongooseTest.cpp:53:pcb       3 0x7 2 0x5555555863b0
127b23a 3 mongoose.c:4233:read_conn     2 0x5 snd 0/2048 rcv 0/2048 n=243 err=4
127b23a 2 MongooseTest.cpp:53:pcb       4 0x8 1 0x5555555863b0
127b23a 3 mongoose.c:4437:mg_mkpipe     4 0x8 pipe 6
127b623 2 MongooseTest.cpp:53:pcb       4 0x8 2 0x5555555863b0
127b623 2 MongooseTest.cpp:53:pcb       3 0x7 2 0x5555555863b0
127ba0a 2 MongooseTest.cpp:53:pcb       4 0x8 2 0x5555555863b0
127ba0a 3 mongoose.c:4233:read_conn     4 0x8 snd 0/0 rcv 0/2048 n=2 err=4
127ba0a 2 MongooseTest.cpp:53:pcb       4 0x8 7 0x5555555863b0
127ba0a 2 MongooseTest.cpp:53:pcb       3 0x7 2 0x5555555863b0
127ba0a 2 MongooseTest.cpp:53:pcb       4 0x8 2 0x5555555863b0
127ba0a 2 MongooseTest.cpp:53:pcb       3 0x7 2 0x5555555863b0
127ba0a 3 mongoose.c:4244:write_conn    2 0x5 snd 65/2048 rcv 0/2048 n=65 err=4
127bdf4 2 MongooseTest.cpp:53:pcb       4 0x8 2 0x5555555863b0
127bdf4 2 MongooseTest.cpp:53:pcb       3 0x7 2 0x5555555863b0
127c1dd 2 MongooseTest.cpp:53:pcb       4 0x8 2 0x5555555863b0
127c1dd 2 MongooseTest.cpp:53:pcb       3 0x7 2 0x5555555863b0
127c5c6 2 MongooseTest.cpp:53:pcb       4 0x8 2 0x5555555863b0
127c5c6 2 MongooseTest.cpp:53:pcb       3 0x7 2 0x5555555863b0
127c9b0 2 MongooseTest.cpp:53:pcb       4 0x8 2 0x5555555863b0
127c9b0 2 MongooseTest.cpp:53:pcb       3 0x7 2 0x5555555863b0
127cd99 2 MongooseTest.cpp:53:pcb       4 0x8 2 0x5555555863b0
127cd99 2 MongooseTest.cpp:53:pcb       3 0x7 2 0x5555555863b0
127d183 2 MongooseTest.cpp:53:pcb       4 0x8 2 0x5555555863b0
127d183 2 MongooseTest.cpp:53:pcb       3 0x7 2 0x5555555863b0
127d56c 2 MongooseTest.cpp:53:pcb       4 0x8 2 0x5555555863b0
127d56c 2 MongooseTest.cpp:53:pcb       3 0x7 2 0x5555555863b0
127d5f0 2 MongooseTest.cpp:53:pcb       4 0x8 2 0x5555555863b0
127d5f0 2 MongooseTest.cpp:53:pcb       3 0x7 2 0x5555555863b0
127d5f0 3 mongoose.c:4233:read_conn     2 0x5 snd 0/2048 rcv 0/2048 n=243 err=4
127d5f0 2 MongooseTest.cpp:53:pcb       5 0x9 1 0x5555555863b0
127d5f0 3 mongoose.c:4437:mg_mkpipe     5 0x9 pipe 6
127d9da 2 MongooseTest.cpp:53:pcb       5 0x9 2 0x5555555863b0
127d9da 2 MongooseTest.cpp:53:pcb       4 0x8 2 0x5555555863b0
127d9da 2 MongooseTest.cpp:53:pcb       3 0x7 2 0x5555555863b0
127ddc1 2 MongooseTest.cpp:53:pcb       5 0x9 2 0x5555555863b0
127ddc1 3 mongoose.c:4233:read_conn     5 0x9 snd 0/0 rcv 0/2048 n=2 err=4
127ddc2 2 MongooseTest.cpp:53:pcb       5 0x9 7 0x5555555863b0
127ddc2 2 MongooseTest.cpp:53:pcb       4 0x8 2 0x5555555863b0
127ddc2 2 MongooseTest.cpp:53:pcb       3 0x7 2 0x5555555863b0
127ddc2 2 MongooseTest.cpp:53:pcb       5 0x9 2 0x5555555863b0
127ddc2 2 MongooseTest.cpp:53:pcb       4 0x8 2 0x5555555863b0
127ddc2 2 MongooseTest.cpp:53:pcb       3 0x7 2 0x5555555863b0
127ddc2 3 mongoose.c:4244:write_conn    2 0x5 snd 65/2048 rcv 0/2048 n=65 err=4
127e1ac 2 MongooseTest.cpp:53:pcb       5 0x9 2 0x5555555863b0
127e1ac 2 MongooseTest.cpp:53:pcb       4 0x8 2 0x5555555863b0
127e1ac 2 MongooseTest.cpp:53:pcb       3 0x7 2 0x5555555863b0

**Comment by scaprile** [comment, marked as answer]
Please read the tutorial to see what events get triggered https://mongoose.ws/tutorials/multi-threaded/
In latest (master branch), we are replacing this by a queue, perhaps you'll feel more comfortable with that https://github.com/cesanta/mongoose/tree/master/examples/multi-threaded

In Linux and Mac, the 7.9 example works as expected, if it does not work for you in your environment, please open an issue and provide all relevant information
Regarding Windows and your compiler, please open an issue and provide all relevant information; it has been tested with MinGW.

**Reply by JplBal** [reply]
Hello 
thank you for the reply.
Effectively, I could be more interested by the queue usage. I will try it
Regards
